docs/04-reference/winglang-spec.md

```diff
@@ -869,30 +869,223 @@ let f = (arg1: num, var arg2: num) => {
+};
+
+assert(my_function(88) == 89);
+assert(my_function(88, 2) == 90);
+```
+
+Alternatively, specifying a default value will also make an argument optional. Using a default value
+in the function declaration ensures that `by` always has a value so there is no need to unwrap it
+(this is why its type is `num` and not `num?`):
+
+```js
+let increment = (x: num, by: num = 1): num {
+  return x + by;
+}
+```
+
+Non-optional arguments can only be used before all optional arguments:
+
+```js
+let my_fun = (a: str, x?: num, y: str): void = { /* ... */ };
+//-----------------------------^^^^^^ ERROR: cannot declare a non-optional argument after an optional
+```
+
+If a function uses a keyword argument struct as the last argument, and there are other optional arguments before,
+it also has to be declared as optional:
+
+```js
+let parse_int = (x: str, radix: num?, opts?: ParseOpts): num { /* ... */ };
+// or
+let parse_int = (x: str, radix: num = 10, opts: ParseOpts = ParseOpts {}): num { /* ... */ };
+```
+
+##### 1.7.1.5 Declaration in function return types
+
+If a function returns an optional type, use the `return nil;` statement to indicate that the value is
+not defined.
+
+> This is currently the only scenario in the language where `nil` is allowed.
+
+```js
+struct Name { first: str, last: str };
+
+let try_parse_name = (full_name: str): Name? => {
+  let parts = full_name.split(" ");
+  if parts.len < 2 {
+    return nil;
+  }
+
+  return Name { first: parts.at(0), last: parts.at(1) };
+}
+```
+
+#### 1.7.2 Testing using `x?`
+
+Object of type `str?` cannot be used within `==` or `!=` expressions because we don't have a null
+value. Instead, use the the `?` operator to get a `bool` that indicates if the optional is defined
+or not:
+
+```js
+let is_address_defined = my_person.address?; // type is `bool`
+
+// or within a condition
+if my_person.address? {
+  print("address is defined but i do not care what it is");
+}
+
+// can be negated
+if !my_person.address? {
+  print("address is not defined");
+}
+```
+
+#### 1.7.3 Unwrapping using `if let` and `let else`
+
+The `if let` statement can be used to test if an optional is defined an *unwrap* it into a
+non-optional variable defined inside the block:
+
+```js
+if let address = my_person.address {
+  print(address.len);
+  print(address); // address is type `str`
+}
+```
+
+> NOTE: `if let` is not the same as `if`. For example, we currently don't support specifying multiple conditions, or unwrapping multiple optionals. This is something we might consider in the future.
+
+The `let .. else` statement can be used to conditionally execute a block if an optional
+does not have a value:
+
+```js
+let address = my_person.address else {
+  throw("address is required");
+};
+
+print("address is ${address}");
+```
+
+#### 1.7.4 Unwrapping or default value using `??`
+
+The `??` operator can be used to unwrap or provide a default value. This returns a value of `T` that can safely be used.
```

Comment:
Note that in TypeScript the type inference for `a ?? b` tries to find the lowest common ancestor between `a` and `b`'s types. So if `a` is `TSuper` and `b` is `TSub` (and `TSub` extends `TSuper`) then the type of the expression will be `TSuper`.
If there's no common ancestor it'll evaluate to a union type of `a` and `b`'s types.
I think in wing we can try to evaluate to the common ancestor, but fail in other cases.

Reply:
Added to roadmap